### PR TITLE
Fixed #26005 -- Convert URIs to IRIs according to RFC

### PR DESCRIPTION
**Pull request by Chronial**
This fixes the broken implementation of 10b17a22bec. As that change touched the wsgi infrastructure someone with understanding of that should review this change.
The original change introduced its `uri_to_iri` function in two places: https://github.com/django/django/commit/10b17a22bec2eaf44c3315614aea87c127caee46#diff-f6d1c75ec606389da5af6558bf57f171R129 and https://github.com/django/django/commit/10b17a22bec2eaf44c3315614aea87c127caee46#diff-97160f50594424a40f2621d5a3c581ccL273.

In the second case the original author clearly didn't want `uri_to_iri` but a fancier `unquote`, so I replaced it with that. In the second case I am not so sure.

Also note that this is in a certain way a backwards incompatible change. Code that assumed that `uri_to_iri` is a fancier `unquote` will break. The new code will also not do what the documentation said before – which is why I had to fix it. But the previous documentation was also self-contradicting, as explained in the trac ticket.

### CODE REVIEW

**Comment by charettes** [comment]
Buildbot, add to whitelist.

File: django/utils/encoding.py [hunk 1]
```diff
@@ -196,20 +193,46 @@ def iri_to_uri(iri):
...
+    else:
+        parts = [bits[0]]
+        append = parts.append
+        global _hextobyte
+        if _hextobyte is None:
+            _allowed_ascii = [45, 46, 95, 126] + list(range(65, 91)) + list(range(97, 123))
+            _hextobyte = {('%02x' % char).encode(): six.int2byte(char)
+                          for char in _allowed_ascii}
+            _hextobyte.update(((a + b).encode(), six.int2byte(int(a + b, 16)))
+                              for a in _hexdig[8:] for b in _hexdig)
```
Comment (charettes) [inline django/utils/encoding.py]: I don't see much point in using a lazily computed global variable here. Can't this be moved to the module level?

Comment (Chronial) [inline django/utils/encoding.py]: This code is based on the [`unquote` implementation in the standardlib](https://github.com/python/cpython/blob/436064bfe71ffd782dc5b942f65b80f7f0d7d2cd/Lib/urllib/parse.py#L554). I couldn't see any disadvantage of this approach, so I followed their example. But now that I'm looking at this again, this approach does not seem completely thread-safe.

The reason given in the standardlib is memory. I don't know the exact memory overhead for the strings and the dict, but I would assume this structure takes less than 4kb of memory.

If you think that's the better approach, I can move the initialization to module level.

Comment (charettes) [inline django/utils/encoding.py]: Agreed this is not thread safe, there's a small race condition window. As the dict is assigned to a global variable on first call anyway (ending up in non-collectable memory) I don't see the point of the lazy initialization.

Comment (Chronial) [inline django/utils/encoding.py]: Moved to module level. Also discovered and fixed another bug with the previous implementation (uppercase hex encodings of ascii letters were not decoded).

File: tests/utils_tests/test_encoding.py [hunk 2]
```diff
@@ -108,6 +108,10 @@ def test_uri_to_iri(self):
             # Valid UTF-8 sequences are decoded.
             ('/%E2%99%A5%E2%99%A5/', '/♥♥/'),
             ('/%E2%99%A5%E2%99%A5/?utf8=%E2%9C%93', '/♥♥/?utf8=✓'),
+            ('/%41%5a%6B/', '/AZk/'),
```
Comment (charettes) [inline tests/utils_tests/test_encoding.py]: I guess this is the extra test for the uppercase bug you just discovered?

Comment (Chronial) [inline tests/utils_tests/test_encoding.py]: Correct.

**Comment by Chronial** [comment]
@timgraham IMHO it would be good if this could make the next release. This is an unavoidable backwards-incompatible change, so the sooner this is done, the smaller the chance that a lot of code depends on the old behavior.

File: django/test/client.py [hunk 3]
```diff
@@ -292,11 +298,12 @@ def _get_path(self, parsed):
...
         if parsed[3]:
             path += str(";") + force_str(parsed[3])
-        path = uri_to_iri(path).encode(UTF_8)
+        path = unquote_to_bytes(path) if six.PY3 else unquote(path)
+        path = repercent_broken_unicode(path)
         # Under Python 3, non-ASCII values in the WSGI environ are arbitrarily
         # decoded with ISO-8859-1. We replicate this behavior here.
         # Refs comment in `get_bytes_from_wsgi()`.
-        return path.decode(ISO_8859_1) if six.PY3 else path
+        return path.decode('iso-8859-1') if six.PY3 else path
```
Comment (timgraham) [inline django/test/client.py]: Is this change needed?

File: django/utils/encoding.py [hunk 4]
```diff
@@ -196,20 +193,41 @@ def iri_to_uri(iri):
     return quote(force_bytes(iri), safe=b"/#%[]=:;$&()+,!?*@'~")
 
 
+_decode_ranges = [[45, 46, 95, 126], range(65, 91), range(97, 123), range(128, 256)]
```
Comment (timgraham) [inline django/utils/encoding.py]: Could you add some explanatory comments for this stuff? I'm not too familiar with the relevant RFCs, etc.

File: django/utils/encoding.py [hunk 5]
```diff
@@ -196,20 +193,41 @@ def iri_to_uri(iri):
...
+_hextobyte = {(fmt % char).encode(): six.int2byte(char)
```
Comment (timgraham) [inline django/utils/encoding.py]: please use this indent style:
```
_hextobyte = {
    (fmt % char).encode(): six.int2byte(char)
    for range in _decode_ranges for char in range for fmt in ['%02x', '%02X']
}
```

File: django/utils/encoding.py [hunk 6]
```diff
@@ -196,20 +193,41 @@ def iri_to_uri(iri):
...
     uri = force_bytes(uri)
-    iri = unquote_to_bytes(uri) if six.PY3 else unquote(uri)
+
+    bits = uri.split(b'%')
+    if len(bits) == 1:
+        iri = uri
+    else:
+        parts = [bits[0]]
+        append = parts.append
+        global _hextobyte
```
Comment (timgraham) [inline django/utils/encoding.py]: drop 'global'?

File: django/utils/encoding.py [hunk 7]
```diff
@@ -196,20 +193,41 @@ def iri_to_uri(iri):
...
+        for item in bits[1:]:
+            try:
+                append(_hextobyte[item[:2]])
+                append(item[2:])
```
Comment (timgraham) [inline django/utils/encoding.py]: I guess the KeyError might happen in the live above so could line go to an else clause of the try/except?

File: docs/ref/unicode.txt [hunk 8]
```diff
@@ -221,19 +221,17 @@ result.
...
-    '%A9helloworld'
+    >>> uri_to_iri('%A9hello%3Fworld')
+    '%A9hello%3Fworld'
 
-In the first example, the UTF-8 characters and reserved characters are
-unquoted. In the second, the percent-encoding remains unchanged because it
-lies outside the valid UTF-8 range.
+In the first example, the UTF-8 characters are unquoted. In the second, the 
+percent-encodings remains unchanged because they lie outside the valid UTF-8 
+range / represent a reserved character.
```
Comment (timgraham) [inline docs/ref/unicode.txt]: does the slash mean "and" or "or"?

**Comment by timgraham** [comment]
@loic, since you merged the original patch, I'm wondering if you have any feedback?

File: django/test/client.py (hunk 3 above)
Comment (Chronial) [inline django/test/client.py]: No, I made it because IMHO this is better style. My understanding is that [these constants](https://github.com/django/django/blob/master/django/core/handlers/wsgi.py#L21) only exist for py2/py3 compatibility. Their use was needed in the `UTF_8` line. The new code only does decoding on py3, so it doesn't need the constants and can just pass the string. This seemed strictly more pythonic to me.

Should I change it back? :)

Comment (timgraham) [inline django/test/client.py]: Not sure, I haven't looked into the history of the constants but I figured they were to prevent typos in the strings.

Comment (Chronial) [inline django/test/client.py]: won't both versions (constant/string) throw an exception on typo?

File: django/utils/encoding.py (hunk 7 above)
Comment (Chronial) [inline django/utils/encoding.py]: This code is from the [cpython `unquote` implementation](https://github.com/python/cpython/blob/436064bfe71ffd782dc5b942f65b80f7f0d7d2cd/Lib/urllib/parse.py#L578-L584). Your point is correct, but I would rather have this be same as in the standart lib.

Your call.

File: django/utils/encoding.py (hunk 4 above)
Comment (Chronial) [inline django/utils/encoding.py]: done.

File: django/utils/encoding.py (hunk 5 above)
Comment (Chronial) [inline django/utils/encoding.py]: done.

File: django/utils/encoding.py (hunk 6 above)
Comment (Chronial) [inline django/utils/encoding.py]: done.

File: docs/ref/unicode.txt (hunk 8 above)
Comment (Chronial) [inline docs/ref/unicode.txt]: clarified.

File: django/utils/encoding.py [hunk 9]
```diff
@@ -196,20 +193,45 @@ def iri_to_uri(iri):
...
+    This is the algorithm from section 3.2 of RFC 3987, excluding step 4.
 
     Takes an URI in ASCII bytes (e.g. '/I%20%E2%99%A5%20Django/') and returns
     unicode containing the encoded result (e.g. '/I \xe2\x99\xa5 Django/').
     """
     if uri is None:
         return uri
     uri = force_bytes(uri)
-    iri = unquote_to_bytes(uri) if six.PY3 else unquote(uri)
+
```
Comment (loic) [inline django/utils/encoding.py]: Is this code based on an existing implementation? If that's the case, we should specify it / link to it.

File: django/utils/encoding.py [hunk 10]
```diff
@@ -196,20 +193,45 @@ def iri_to_uri(iri):
...
+}
+
+
 def uri_to_iri(uri):
     """
     Converts a Uniform Resource Identifier(URI) into an Internationalized
     Resource Identifier(IRI).
 
-    This is the algorithm from section 3.2 of RFC 3987.
+    This is the algorithm from section 3.2 of RFC 3987, excluding step 4.
```
Comment (loic) [inline django/utils/encoding.py]: Step 4 is:
> Re-percent-encode all octets produced in step 3 that in UTF-8 represent characters that are not appropriate according to sections.

Isn't that what we do when leaving broken sequences unchanged?

Comment (Chronial) [inline django/utils/encoding.py]: No, that is step 3:
> `3.` Re-percent-encode any octet produced in step 2 that is not part of a strictly legal UTF-8 octet sequence.

I was confused by step 4 at first too, but it is clarified in the referenced sections:
> `4.` Re-percent-encode all octets produced in step 3 that in UTF-8 represent characters that are not appropriate according to sections 2.2, 4.1, and 6.1.

Section 2.2 says:
> […] except that the range of unreserved characters is expanded to include UCS characters, *with the restriction that private UCS characters can occur only in query parts*

Section 4.1 says:
> IRIs MUST NOT contain bidirectional formatting characters (LRM, RLM, LRE, RLE, LRO, RLO, and PDF).

Section 6.1 says:
> The UCS contains many areas of characters for which there are strong visual look-alikes.  Because of the likelihood of        transcription errors, these also should be avoided.  This        includes the full-width equivalents of Latin characters,        half-width Katakana characters for Japanese, and many others.  It        also includes many look-alikes of "space", "delims", and       "unwise", characters excluded in [RFC3491].

We check none of the above.

File: django/utils/encoding.py (hunk 9 above)
Comment (Chronial) [inline django/utils/encoding.py]: Yes, this is a modification of [the `unquote_to_bytes()` function from the standardlib](https://github.com/python/cpython/blob/9043c80b0daaddb5bff1e37432d503910506727f/Lib/urllib/parse.py#L557-L585), that was called here before.

Should I note this in a comment?

Comment (loic) [inline django/utils/encoding.py]: Yep

Comment (loic) [inline django/utils/encoding.py]: Also add in the comment what modification was needed.

File: django/utils/encoding.py (hunk 10 above)
Comment (loic) [inline django/utils/encoding.py]: Interesting! I guess we could implement it at some point, but that seems like a fair amount of work.

File: django/utils/encoding.py (hunk 9 above)
Comment (Chronial) [inline django/utils/encoding.py]: That does not seem very practical. This code is far enough removed from the original, that it is a lot easier to just look at it and see what it does than to understand how it differs from the original. The only concept that is the same is “split the string at "%" and then look up the two characters after the "%" in a map”. But even *why* this lookup is done is not the same anymore.

Comment (loic) [inline django/utils/encoding.py]: Fair enough, I assumed it was a small tweak to the stdlib version.

File: django/utils/encoding.py (hunk 10 above)
Comment (Chronial) [inline django/utils/encoding.py]: IMHO, that is actually not that difficult. `2.2` and `4.1` are quite obvious and `6.1` references `RFC3491`, which just says:
>    This profile specifies prohibiting using the following tables from   [STRINGPREP]:
>
>   Table C.1.2
   Table C.2.2
   Table C.3
   Table C.4
   Table C.5
   Table C.6
   Table C.7
   Table C.8
   Table C.9

And it just so happens that `stringprep` is already part of the standardlib: https://docs.python.org/2/library/stringprep.html :).

The more relevant question is whether we actually want this. `uri_to_iri` is used to transform the page url for routing etc. This seems rather unwise to me, but I don't understand that infrastructure enough to want to touch it.

If we implement step 4, people need to understand step 4 in order to use django routing and make sense of `request.path`. Do we want that?

Comment (timgraham) [inline django/utils/encoding.py]: Loic, reminder if you can respond.

**Comment by timgraham** [comment]
@Chronial, would you like to continue this? Things might be simpler now that master supports Python 3 only. If there are outstanding questions, I'll try to learn enough to give a response if no one else can.

**Comment by Chronial** [comment]
@timgraham yes, I would still be very happy to work towards getting this merged.

**Comment by timgraham** [comment]
Yes, please rebase and let me know if you need some question answered.

**Comment by Chronial** [comment]
@timgraham done. This actually got quite a bit clearer, since 48c34f3336cbbc90606663 implemented the correct fix for what the broken 10b17a2 was attempting to do.

File: django/utils/encoding.py [hunk 11]
```diff
@@ -151,20 +151,54 @@ def iri_to_uri(iri):
...
 
     Takes an URI in ASCII bytes (e.g. '/I%20%E2%99%A5%20Django/') and returns
     a string containing the encoded result (e.g. '/I \xe2\x99\xa5 Django/').
     """
     if uri is None:
         return uri
     uri = force_bytes(uri)
-    iri = unquote_to_bytes(uri)
+
+    bits = uri.split(b'%')
```
Comment (timgraham) [inline django/utils/encoding.py]: Could some incline comments help understanding these lines?

File: django/test/client.py (hunk 3 above)
Comment (timgraham) [inline django/test/client.py]: Thanks, this inspired a separate cleanup: https://github.com/django/django/pull/8032.

**Comment by Chronial** [comment]
@timgraham Applied the requested changes.

**Comment by timgraham** [comment]
merged in 03281d8fe7a32f580a85235659d4fbb143eeb867, thanks!

**Comment by Chronial** [comment]
Thanks :)
